app/src/main/java/com/sdp13epfl2021/projmag/database/FirebaseFileDatabase.kt

```diff
@@ -1,103 +1,133 @@
+
+
+    override fun getFile(
+        fileUrl: String,
+        destinationFolder: File,
+        onSuccess: (File) -> Unit,
+        onFailure: (Exception) -> Unit
+    ) {
+        val fileRef: StorageReference
+        val destinationFile: File
+
+        try {
+            if (!destinationFolder.mkdirs() && !destinationFolder.isDirectory) {
+                throw IOException("The destination folder \"${destinationFolder.path}\" can't be created.")
+            }
+            fileRef = storage.getReferenceFromUrl(fileUrl)
+            destinationFile = File(destinationFolder, fileRef.name)
+        } catch (e: Exception) {
+            GlobalScope.launch { onFailure(e) }
+            return
+        }
+
+        if (destinationFile.exists()) {
+            GlobalScope.launch { onSuccess(destinationFile) }
+            return
+        }
+
+        fileRef
+            .getFile(destinationFile)
+            .addOnSuccessListener { onSuccess(destinationFile) }
+            .addOnFailureListener(onFailure)
+    }
+
+
+    override fun pushFile(
+        file: File,
+        onSuccess: (Uri) -> Unit,
+        onFailure: (Exception) -> Unit
+    ) {
+        if (!file.exists() || !file.isFile) {
+            GlobalScope.launch { onFailure(IOException("The file \"${file.path}\" can't be read.")) }
+            return
+        }
+
+        val user = auth.currentUser
+        if (user == null) {
+            GlobalScope.launch { onFailure(SecurityException("Pushing file can only be done by authenticated user.")) }
+            return
+        }
+
+        val fileRef = rootRef
+            .child(user.uid)
+            .child("${UUID.randomUUID()}_${file.name}")
+
+        fileRef
+            .putFile(file.toUri())
+            .continueWithTask { task ->
+                if (!task.isSuccessful) {
+                    task.exception?.let {
+                        throw it
+                    }
+                }
+                fileRef.downloadUrl
+            }
+            .addOnSuccessListener(onSuccess)
+            .addOnFailureListener(onFailure)
+    }
+
+    override fun pushFileFromUri(
+        uri: Uri,
+        onSuccess: (Uri) -> Unit,
+        onFailure: (Exception) -> Unit
+    ) {
+
+        val user = auth.currentUser
+        if (user == null) {
+            GlobalScope.launch { onFailure(SecurityException("Pushing file can only be done by authenticated user.")) }
+            return
+        }
+
+        val fileRef = rootRef
+            .child(user.uid)
+            .child("${UUID.randomUUID()}_${uri.toString()}")
+
+        fileRef
+            .putFile(uri)
+            .continueWithTask { task ->
+                if (!task.isSuccessful) {
+                    task.exception?.let {
+                        throw it
+                    }
+                }
+                fileRef.downloadUrl
+            }
+            .addOnSuccessListener(onSuccess)
+            .addOnFailureListener(onFailure)
+    }
```

Comment:
Since both functions use putFile with a Uri, maybe we could refactor this to avoid code duplication.

Reply:
It has been factored out in [this commit](https://github.com/asfoury/projmag/pull/44/commits/696c1d9956453fa5016f01ca3e29d6575041cd79)